src/pages/node-express/lesson4-a1.mdx

```diff
@@ -0,0 +1,103 @@
+
+```js
+app.get('/', logger, (req, res) => {
+    ...
+})
+```
+
+This means the `logger` middleware function will run before any `GET` request to the `/` path.
+The second way to invoke middleware is via an `app.use()` statement:
+
+```js
+app.use(["/path1", "/path2"], logger);
+```
+
+In this case, the first argument is a path or an array of paths indicating the urls for which the middleware is called. This argument is optional. If you leave it out, it is called for _**all**_ urls. When using an `app.use()` statement, order is important! For example, if you put `app.use(express.json())` after your `app.post()` statement, the `app.post()` won't work as expected, because the body will not have been parsed into JSON yet. Call your logger using the first method, in one of your `app.get()` statements, and verify that it works. Then, take the logger call out of your `app.get()` statement, and call it via `app.use()`, for all paths, instead. Verify that it still works.
+
+Next, you need to implement some APIs for people. You have a require statement for `./data.js` that gets the value of products. Get the value for people, also from `./data.js` (add this in the same require statement). Then implement an `app.get` for `/api/v1/people`. Test it with your browser.  You are returning JSON, so you call res.json(...) to send the data back. You now need to implement an `app.post` for `/api/v1/people`. This is to add an entry to the people array. Post operations are sent from the browser with a "request body". You need to add middleware to parse this body into a Javascript object. The following statements do this parsing, returning the result as a hash in `req.body`.
+
+```js
+app.use(express.urlencoded({ extended: false }));
+app.use(express.json());
+```
+
+The first of these parses url-encoded data, which is the format that is sent by an HTML form. This is not typically needed if you are only implementing a JSON API. The second statement parses a JSON body. You need these statements before your `app.post()` statement, so that the body is parsed before you do the rest of the processing. Now you implement the `app.post` statement for `/api/v1/people`. The statement should check `req.body` to see if there is a `req.body.name` property. If not, it should return JSON for an error, as follows:
+
+```js
+res.status(400).json({ success: false, message: "Please provide a name" });
+```
+
+This sets the HTTP result code to [400](https://http.dev/400), which means there was an error on the client side, and also returns an error message. But suppose there is a value in req.body.name. You want to add this entry to the people array, as follows:
```

Comment:
added links where status codes are first mentioned to `http.dev` for further reading